File: src/ios/LukuIDSDK/Tests/LukuIDSDKTests/GuardCardParityTests.swift
// SPDX-License-Identifier: Apache-2.0
import XCTest
@testable import LukuIDSDK

final class LukuIDEnvironmentParityTests: XCTestCase {

    private func getValidEnvelope() -> [String: Any] {
        let fileManager = FileManager.default
        var currentURL = URL(fileURLWithPath: #file)
        var sampleURL: URL?

        while currentURL.path != "/" {
            let candidate = currentURL.appendingPathComponent("samples/envelopes/dev/1.0.0/valid_envelope.json")
            if fileManager.fileExists(atPath: candidate.path) {
                sampleURL = candidate
                break
            }
            currentURL = currentURL.deletingLastPathComponent()
        }

        guard let url = sampleURL, let data = try? Data(contentsOf: url) else {
            fatalError("Could not locate valid_envelope.json")
        }

        return try! JSONSerialization.jsonObject(with: data, options: []) as! [String: Any]
    }

    private func verify(_ envelope: [String: Any]) -> [VerificationIssue] {
        LukuFile.verifyEnvelope(
            envelope: envelope,
            options: LukuVerifyOptions(
                allowUntrustedRoots: false,
                skipCertificateTemporalChecks: true,
                trustProfile: "dev"
            )
        )
    }

    func testMissingDetachedDacSignatureFails() {
        var envelope = getValidEnvelope()
        envelope.removeValue(forKey: "attestation_dac_signature")
        envelope.removeValue(forKey: "attestation_signature")
        if var identity = envelope["identity"] as? [String: Any] {
            identity.removeValue(forKey: "dac_signature")
            identity.removeValue(forKey: "signature")
            envelope["identity"] = identity
        }

        let issues = verify(envelope)
        XCTAssertTrue(
            issues.contains { $0.code == "ATTESTATION_FAILED" && $0.message.contains("attestationSig missing") },
            "Expected attestationSig missing failure, got \(issues)"
        )
    }

    func testHeartbeatSignatureRequiresTrustedHeartbeatTimestamp() {
        var envelope = getValidEnvelope()
        guard var identity = envelope["identity"] as? [String: Any] else {
            XCTFail("Missing identity")
            return
        }

        // heartbeat_signature is already present in the valid envelope
        identity.removeValue(forKey: "last_sync_utc")
        envelope["identity"] = identity
        envelope.removeValue(forKey: "last_sync_utc")

        let issues = verify(envelope)
        XCTAssertTrue(
            issues.contains { $0.code == "ATTESTATION_FAILED" && $0.message.contains("Missing trusted heartbeat timestamp") },
            "Expected missing trusted heartbeat timestamp failure, got \(issues)"
        )
    }

    func testHeartbeatSignatureMustMatchHeartbeatPayload() {
        var envelope = getValidEnvelope()
        guard var identity = envelope["identity"] as? [String: Any] else {
            XCTFail("Missing identity")
            return
        }

        // heartbeat_signature is already present
        identity["last_sync_utc"] = 1777286310
        envelope["identity"] = identity

        let issues = verify(envelope)
        XCTAssertTrue(
            issues.contains { $0.code == "ATTESTATION_FAILED" && $0.message.contains("SLAC (heartbeat)") },
            "Expected heartbeat payload verification failure, got \(issues)"
        )
    }

    func testLastSyncCannotBeAfterRecordTimestamp() {
        var envelope = getValidEnvelope()
        guard var identity = envelope["identity"] as? [String: Any] else {
            XCTFail("Missing identity")
            return
        }

        // heartbeat_signature is already present
        identity["last_sync_utc"] = 1781119207
        envelope["identity"] = identity

        let issues = verify(envelope)
        XCTAssertTrue(
            issues.contains { $0.code == "LAST_SYNC_AFTER_RECORD" },
            "Expected LAST_SYNC_AFTER_RECORD, got \(issues)"
        )
    }
}

File: src/rust/lukuid-sdk/tests/guardcard_parity_tests.rs
use lukuid_sdk::{LukuFile, LukuVerifyOptions};
use serde_json::Value;
use std::fs;
use std::path::PathBuf;

fn resolve_sample() -> PathBuf {
    let mut current = std::env::current_dir().unwrap();
    loop {
        let candidate = current
            .join("samples")
            .join("envelopes")
            .join("dev")
            .join("1.0.0")
            .join("valid_envelope.json");
        if candidate.exists() {
            return candidate;
        }
        if !current.pop() {
            break;
        }
    }
    PathBuf::from("../../../samples/envelopes/dev/1.0.0/valid_envelope.json")
}

fn get_valid_envelope() -> Value {
    let path = resolve_sample();
    let content = fs::read_to_string(&path).expect("Failed to read valid_envelope.json");
    serde_json::from_str(&content).expect("Failed to parse valid_envelope.json")
}

fn verify(envelope: &Value) -> Vec<lukuid_sdk::VerificationIssue> {
    let mut options = LukuVerifyOptions::default();
    options.allow_untrusted_roots = false;
    options.skip_certificate_temporal_checks = true;
    options.trust_profile = "dev".to_string();
    LukuFile::verify_envelope(envelope, options)
}

#[test]
fn test_missing_detached_dac_signature_fails() {
    let mut envelope = get_valid_envelope();
    if let Some(object) = envelope.as_object_mut() {
        object.remove("attestation_signature");
    }
    if let Some(identity) = envelope.get_mut("identity").and_then(|v| v.as_object_mut()) {
        identity.remove("signature");
    }

    let issues = verify(&envelope);
    assert!(
        issues.iter().any(|issue| issue.code == "ATTESTATION_FAILED" && issue.message.contains("attestationSig missing")),
        "Expected attestationSig missing failure, got: {:?}",
        issues
    );
}

#[test]
fn test_heartbeat_signature_requires_trusted_heartbeat_timestamp() {
    let mut envelope = get_valid_envelope();
    let signature = envelope["identity"]["signature"].as_str().unwrap().to_string();
    envelope["heartbeat_signature"] = Value::String(signature);
    if let Some(identity) = envelope.get_mut("identity").and_then(|v| v.as_object_mut()) {
        identity.remove("last_sync_utc");
    }
    if let Some(object) = envelope.as_object_mut() {
        object.remove("last_sync_utc");
    }

    let issues = verify(&envelope);
    assert!(
        issues.iter().any(|i| i.code == "ATTESTATION_FAILED" && i.message.contains("Missing trusted heartbeat timestamp")),
        "Expected missing trusted heartbeat timestamp failure, got: {:?}",
        issues
    );
}

#[test]
fn test_heartbeat_signature_must_match_heartbeat_payload() {
    let mut envelope = get_valid_envelope();
    let signature = envelope["identity"]["signature"].as_str().unwrap().to_string();
    envelope["heartbeat_signature"] = Value::String(signature);
    if let Some(identity) = envelope.get_mut("identity").and_then(|v| v.as_object_mut()) {
        identity.insert("last_sync_utc".to_string(), Value::from(1777286310_i64));
    }

    let issues = verify(&envelope);
    assert!(
        issues.iter().any(|i| i.code == "ATTESTATION_FAILED" && i.message.contains("SLAC (heartbeat)")),
        "Expected heartbeat signature verification failure, got: {:?}",
        issues
    );
}

#[test]
fn test_last_sync_cannot_be_after_record_timestamp() {
    let mut envelope = get_valid_envelope();
    let signature = envelope["identity"]["signature"].as_str().unwrap().to_string();
    envelope["heartbeat_signature"] = Value::String(signature);
    if let Some(identity) = envelope.get_mut("identity").and_then(|v| v.as_object_mut()) {
        identity.insert("last_sync_utc".to_string(), Value::from(1777286312_i64));
    }

    let issues = verify(&envelope);
    assert!(
        issues.iter().any(|i| i.code == "LAST_SYNC_AFTER_RECORD"),
        "Expected LAST_SYNC_AFTER_RECORD, got: {:?}",
        issues
    );
}
